File: backend/src/main/java/br/com/calciolari/datahub/imports/domain/hints/IncompleteDate.java
package br.com.calciolari.datahub.imports.domain.hints;

import java.util.Optional;
import java.util.OptionalInt;

/**
 * Day/month extracted from a filename when the year is absent or incomplete.
 * Never filled with the JVM "current" year.
 */
public record IncompleteDate(int day, int month, OptionalInt year) {

	public IncompleteDate(int day, int month) {
		this(day, month, OptionalInt.empty());
	}

	public IncompleteDate {
		if (day < 1 || day > 31) {
			throw new IllegalArgumentException("day out of range: " + day);
		}
		if (month < 1 || month > 12) {
			throw new IllegalArgumentException("month out of range: " + month);
		}
		year = year == null ? OptionalInt.empty() : year;
	}

	public boolean hasYear() {
		return year.isPresent();
	}
}

File: backend/src/Calciolari.DataHub/Imports/Domain/Hints/IncompleteDate.cs
namespace Calciolari.DataHub.Imports.Domain.Hints;

/// <summary>
/// Day/month extracted from a filename when the year is absent or incomplete.
/// Never filled with the clock's current year.
/// </summary>
public sealed record IncompleteDate(int Day, int Month, int? Year = null)
{
    public bool HasYear => Year is not null;

    public static IncompleteDate Create(int day, int month, int? year = null)
    {
        if (day < 1 || day > 31)
        {
            throw new ArgumentOutOfRangeException(nameof(day), "day out of range: " + day);
        }

        if (month < 1 || month > 12)
        {
            throw new ArgumentOutOfRangeException(nameof(month), "month out of range: " + month);
        }

        return new IncompleteDate(day, month, year);
    }
}
